### ISSUE TITLE
"pipenv lock -r " doesn't output extra index information

### ISSUE THREAD

**Issue by jakul** [pypa/pipenv#1987]
When the pipfile includes private PyPi's, `pipenv lock -r` should output the index url with the requirement.

Here's an example:

```
craig@amogato2:~/src/boughtbymany/private_pypi_lock
$ pipenv lock -r
click==6.7
my-python-project==0.1.0

craig@amogato2:~/src/boughtbymany/private_pypi_lock
$ cat Pipfile

craig@amogato2:~/src/boughtbymany/private_pypi_lock
$ pipenv run pip install my-python-project==0.1.0
Collecting my-python-project==0.1.0
  Could not find a version that satisfies the requirement my-python-project==0.1.0 (from versions: )
No matching distribution found for my-python-project==0.1.0

craig@amogato2:~/src/boughtbymany/private_pypi_lock
$ cat Pipfile
[[source]]
url = "https://pypi.python.org/simple"
verify_ssl = true
name = "pypi"

[[source]]
url = "https://testpypi.python.org/simple"
verify_ssl = true
name = "test_pypi"

[packages]
my_python_project = {version = "==0.1.0", index = "test_pypi"}

[dev-packages]

[requires]
python_version = "2.7"
```

<details><summary>$ python -m pipenv.help output</summary>

Pipenv version: `'11.10.0'`

Pipenv location: `'/usr/local/Cellar/pipenv/11.10.0/libexec/lib/python3.6/site-packages/pipenv'`

Python location: `'/usr/local/Cellar/pipenv/11.10.0/libexec/bin/python3.6'`

Other Python installations in `PATH`:

  - `2.7`: `/usr/bin/python2.7`
  - `2.7`: `/usr/bin/python2.7`
  - `3.6`: `/usr/local/bin/python3.6m`
  - `3.6`: `/usr/local/bin/python3.6`
  - `3.6`: `/usr/local/bin/python3.6`

  - `2.7.10`: `/usr/bin/python`
  - `3.6.5`: `/usr/local/bin/python3`
  - `3.6.5`: `/usr/local/bin/python3`

PEP 508 Information:

```
{'implementation_name': 'cpython',
 'implementation_version': '3.6.5',
 'os_name': 'posix',
 'platform_machine': 'x86_64',
 'platform_python_implementation': 'CPython',
 'platform_release': '17.5.0',
 'platform_system': 'Darwin',
 'platform_version': 'Darwin Kernel Version 17.5.0: Mon Mar  5 22:24:32 PST '
                     '2018; root:xnu-4570.51.1~1/RELEASE_X86_64',
 'python_full_version': '3.6.5',
 'python_version': '3.6',
 'sys_platform': 'darwin'}
```

System environment variables:

  - `LDFLAGS`
  - `VIRTUALENVWRAPPER_SCRIPT`
  - `VIRTUALENVWRAPPER_PROJECT_FILENAME`
  - `TERM_PROGRAM`
  - `NVM_CD_FLAGS`
  - `TERM`
  - `SHELL`
  - `HISTSIZE`
  - `CPPFLAGS`
  - `TMPDIR`
  - `Apple_PubSub_Socket_Render`
  - `TERM_PROGRAM_VERSION`
  - `TERM_SESSION_ID`
  - `LC_ALL`
  - `NVM_DIR`
  - `USER`
  - `CROSS_ACCOUNT_STACK_AWS_PROFILE_NAME_BBM_MASTER`
  - `COMMAND_MODE`
  - `SSH_AUTH_SOCK`
  - `__CF_USER_TEXT_ENCODING`
  - `FAB_COMPLETION_CACHE_TASKS`
  - `WORKON_HOME`
  - `PROJECT_HOME`
  - `VIRTUALENVWRAPPER_PYTHON`
  - `NPM_TOKEN`
  - `PATH`
  - `WERKZEUG_DEBUG_PIN`
  - `VIRTUALENVWRAPPER_HOOK_DIR`
  - `PWD`
  - `LANG`
  - `ITERM_PROFILE`
  - `XPC_FLAGS`
  - `FAB_COMPLETION_CACHED_TASKS_FILENAME`
  - `PYTEST_ADDOPTS`
  - `HISTCONTROL`
  - `XPC_SERVICE_NAME`
  - `SHLVL`
  - `HOME`
  - `COLORFGBG`
  - `CFLAGS`
  - `ITERM_SESSION_ID`
  - `LOGNAME`
  - `LC_CTYPE`
  - `PKG_CONFIG_PATH`
  - `PROMPT_COMMAND`
  - `VIRTUALENVWRAPPER_WORKON_CD`
  - `EXTRA_CFLAGS`
  - `SECURITYSESSIONID`
  - `EXTRA_LDFLAGS`
  - `COLORTERM`
  - `OLDPWD`
  - `_`
  - `PYTHONDONTWRITEBYTECODE`
  - `PIP_PYTHON_PATH`
  - `PYTHONUNBUFFERED`

Pipenv–specific environment variables:

Debug–specific environment variables:

  - `PATH`: `/Users/craig/.yarn/bin:/Users/craig/.local/bin:/usr/local/opt/openssl/bin:/usr/local/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin`
  - `SHELL`: `/bin/bash`
  - `LANG`: `en_GB.UTF-8`
  - `PWD`: `/Users/craig/src/boughtbymany/private_pypi_lock`

---------------------------

Contents of `Pipfile` ('/Users/craig/src/boughtbymany/private_pypi_lock/Pipfile'):

```toml
[[source]]
url = "https://pypi.python.org/simple"
verify_ssl = true
name = "pypi"

[[source]]
url = "https://testpypi.python.org/simple"
verify_ssl = true
name = "test_pypi"

[packages]
#restup = {version = "==0.0.1", index = "test_pypi"}
my_python_project = {version = "==0.1.0", index = "test_pypi"}

[dev-packages]

[requires]
python_version = "2.7"


```

Contents of `Pipfile.lock` ('/Users/craig/src/boughtbymany/private_pypi_lock/Pipfile.lock'):

```json
{
    "_meta": {
        "hash": {
            "sha256": "f87225d09fae469b08112722b9b8a8b7edbe5be62b03bf59357e56da67d5c648"
        },
        "pipfile-spec": 6,
        "requires": {
            "python_version": "2.7"
        },
        "sources": [
            {
                "name": "pypi",
                "url": "https://pypi.python.org/simple",
                "verify_ssl": true
            },
            {
                "name": "test_pypi",
                "url": "https://testpypi.python.org/simple",
                "verify_ssl": true
            }
        ]
    },
    "default": {
        "click": {
            "hashes": [
                "sha256:29f99fc6125fbc931b758dc053b3114e55c77a6e4c6c3a2674a2dc986016381d",
                "sha256:f15516df478d5a56180fbf80e68f206010e6d160fc39fa508b65e035fd75130b"
            ],
            "version": "==6.7"
        },
        "my-python-project": {
            "hashes": [
                "sha256:5d552f5416681be4dd7df70eff29ff76d5c4dc6078615311ca735d0c3161865c",
                "sha256:a5d7499232f15fcf313d7466ec1b0cf048b0fc0063c179a37ac4cb8d2a447839"
            ],
            "index": "test_pypi",
            "version": "==0.1.0"
        }
    },
    "develop": {}
}

```
</details>

------------

##### Expected result
```
$ pipenv lock -r
click==6.7
my-python-project==0.1.0 --extra-index-url https://testpypi.python.org/simple
```
##### Actual result
```
$ pipenv lock -r
click==6.7
my-python-project==0.1.0
```
##### Steps to replicate

1) Use Pipfile above
2) Run `pipenv lock -r`

**Comment by jakul** [comment]
@techalchemy This issue is related to the work you have been doing in https://github.com/pypa/pipenv/pull/1980

**Comment by techalchemy** [comment]
Try `pip install -e git+https://github.com/pypa/pipenv.git@1973-extra-index-urls#egg=pipenv` and let me know if it works

**Comment by jakul** [comment]
No luck:

```
craig@amogato2:~/src/boughtbymany/private_pypi_lock
$ pipenv lock -r
click==6.7
my-python-project==0.1.0
```


And just to confirm I've got your code branch:
```
$ cat /Users/craig/.local/venvs/pipenv/src/pipenv/pipenv/core.py | grep valid_indexes
        valid_indexes = []
            valid_indexes = [p['name'] for p in project.parsed_pipfile['source']]
        if not is_valid_url(index) and index in valid_indexes:
```

**Comment by techalchemy** [comment]
fix is up, same branch, same command, let me know this time

**Comment by jakul** [comment]
That seems to have worked better:

```
$ pipenv lock -r
-i https://pypi.python.org/simple
--extra-index-url https://testpypi.python.org/simple
click==6.7
my-python-project==0.1.0
```


I still don't think this is the complete solution though, as my Pipfile specified that I want to use `https://testpypi.python.org/simple` for _only_ `my_python_project`, not for all requirements. If there is a requirement which exists in both the public PyPi and the private PyPi, then `pip` may not install the correct version

**Comment by techalchemy** [comment]
@jakul that should only be the case if both indexes offer the same version but one implementation is different. That is not a supported use case at the moment

For outputting to requirements files I don’t know that the required syntax would be allowed, I think it breaks.

**Comment by techalchemy** [comment]
Closed by #1980
